### PR TITLE
PERF: period factorization

### LINKED ISSUE #14338: PERF: Period factorization very slow in 0.19.0
```python
df = DataFrame(data={'data': np.random.randint(0, 100, size=5500000),
                         'date': [dt(2016, 1, 1)] * 5500000})
for period, g in df.groupby(pd.DatetimeIndex(df.date).to_period('D')):
    print(g)
    
```

#### Expected Output
outputs dataframe

#### Output of ``pd.show_versions()``
0.19.0

The output is not the issue, the issue is that in any version before 0.19.0, this was incredibly fast, like ~1 second or less. With 0.19.0, after waiting many minutes I just give up.

### PR DESCRIPTION
**Pull request by chris-b1**
Merged in #14419

---
- [x] closes #14338
- [x] tests not needed / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

asv 

```
   before     after       ratio
  [c41c6511] [96b364a4]
-     2.44s    46.28ms      0.02  groupby.groupby_period.time_groupby_sum
```

cc @sinhrks , @bmoscon

### CODE REVIEW

File: asv_bench/benchmarks/groupby.py [hunk 1]
```diff
@@ -548,6 +548,19 @@ def time_groupby_sum(self):
         self.df.groupby(['a'])['b'].sum()
 
 
+class groupby_period(object):
+    # GH 14338
+    goal_time = 0.2
+
+    def setup(self):
```
Comment (jreback) [inline asv_bench/benchmarks/groupby.py]: can you add similar benches for dti and dti w/tz

**Review by jreback — CHANGES_REQUESTED** [review]

File: asv_bench/benchmarks/groupby.py (hunk 1 above)
Comment (jreback) [inline asv_bench/benchmarks/groupby.py]: ideally inherit from a common class

File: pandas/core/algorithms.py [hunk 2]
```diff
@@ -285,15 +285,17 @@ def factorize(values, sort=False, order=None, na_sentinel=-1, size_hint=None):
...
-    is_datetimetz_type = is_datetimetz(values)
-    if is_datetimetz_type:
+    if is_datetimetz(values):
         values = DatetimeIndex(values)
-        vals = values.asi8
+
+    if is_period_dtype(values):
+        values = PeriodIndex(values)
+        # period array interface goes to object so intercept
+        vals = values.view(np.int64)
```
Comment (jreback) [inline pandas/core/algorithms.py]: use ``values.asi8``

File: pandas/core/algorithms.py [hunk 3]
```diff
@@ -285,15 +285,17 @@ def factorize(values, sort=False, order=None, na_sentinel=-1, size_hint=None):
...
-    vals = np.asarray(values)
-
-    # localize to UTC
-    is_datetimetz_type = is_datetimetz(values)
-    if is_datetimetz_type:
+    if is_datetimetz(values):
         values = DatetimeIndex(values)
-        vals = values.asi8
+
+    if is_period_dtype(values):
```
Comment (jreback) [inline pandas/core/algorithms.py]: make an elif

Comment (jreback) [inline pandas/core/algorithms.py]: maybe add a comment here about what is going on (helpful to future readers)

**Comment by bmoscon** [comment]
thanks for taking care of this. It looks like this  is targeted for 0.19.1; any rough idea when that will be released?

**Comment by jorisvandenbossche** [comment]
Looks good to me. We only should include a whatsnew notice, depending on the outcome of https://github.com/pydata/pandas/issues/14352

**Comment by jorisvandenbossche** [comment]
Looks good to me! @jreback

**Comment by jorisvandenbossche** [comment]
@chris-b1 whoops, sorry, closed this and I don't see a reopen button ..

**Comment by chris-b1** [comment]
Strange, I can't reopen either, will just make a new PR

**Comment by chris-b1** [comment]
#14419

**Comment by jorisvandenbossche** [comment]
Possibly some glitches due to the repo transfer. Older PRs are also stuck on "Checking for ability to merge automatically…" (that was the reason I wanted to close/reopen this one to retrigger it)

**Pull request closed without merging**